A faster copyto_unaliased!

N5N3 opened the pull request:
~At present, `a .= b` falls to `copyto!(a, b)` if `axes(a) == axes(b)` for performance optimization.
However, as mentioned in #40962 and #39345, the performance of `copyto!(a, b)` is very low in many cases. Thus, I think we can implement a faster version here, based on `@simd`, to solve the problem to some extent.
(I tried to to optimize `copyto_unalias!`, but many generated codes are not vectorized without `@simd`.)~
This PR tries to speed up `copyto_unaliased!` in many cases:
1. **Linear-to-Linear copy** is using `Base.OneTo` as the iterartor instead of `LinearIndices` for faster speed.
2. `zip()` based iterator is used to speed up the most general case.
3. The innermost loop for `CartesianIndices` is separated mannully like `@simd` (but add no `Expr(:loopinfo`) to speed up if the 1st dim's length is larger than 16.
```julia
size:(16, 2048) -> size:(16, 2048)
Car -> Car : before: 17.5μs -> after: 8.5μs
Car -> Lin : before: 52.2μs -> after: 8.5μs
Lin -> Car : before: 75.5μs -> after: 8.77μs
Lin -> Lin : before: 10.6μs -> after: 8.3μs
size:(16, 4, 512) -> size:(16, 4, 512)
Car -> Car : before: 18μs -> after: 9.5μs
Car -> Lin : before: 50.1μs -> after: 9.3μs
Lin -> Car : before: 82μs -> after: 11.6μs
Lin -> Lin : before: 10.7μs -> after: 8.37μs
size:(16, 4, 4, 128) -> size:(16, 4, 4, 128)
Car -> Car : before: 19.7μs -> after: 9.4μs
Car -> Lin : before: 79.3μs -> after: 9.2μs
Lin -> Car : before: 95.8μs -> after: 9.2μs
Lin -> Lin : before: 10.6μs -> after: 8.37μs
size:(16, 4, 4, 4, 32) -> size:(16, 4, 4, 4, 32)
Car -> Car : before: 20.4μs -> after: 9.8μs
Car -> Lin : before: 94.9μs -> after: 9.1μs
Lin -> Car : before: 119μs -> after: 10μs
Lin -> Lin : before: 10.6μs -> after: 8.37μs
```
The above Benchmarks, using `SubArray{Float64}`, are done in 1.7.0-beta3, where the performance regression on `CartesianIndices`'s iteration(#38073) seems to be fixed.

N5N3 committed:
Exclude the slow branch in Broadcast's copyto!

julia> a = randn(50,50); b = randn(50,50); # 2D case
# before
julia> @btime $a[1:end,1:end] .= $b;
  4.900 μs (0 allocations: 0 bytes)
julia> @btime $a[:,:] .= $b;
  721.642 ns (0 allocations: 0 bytes)
# after
julia> @btime $a[1:end,1:end] .= $b;
  355.238 ns (0 allocations: 0 bytes)
julia> @btime $a[:,:] .= $b;
  351.643 ns (0 allocations: 0 bytes)

julia> a = randn(50*50); b = randn(50*50); #1D case
# before 
julia> @btime $a .= $b;
  406.566 ns (0 allocations: 0 bytes)
# after
julia> @btime $a .= $b;
  270.607 ns (0 allocations: 0 bytes)

N5N3 commented:
The failure in mpfr seems unrelated？

oscardssmith commented:
Can you add a 3d benchmark? Other than that, this looks good.

N5N3 commented:
It seems the speed of the general broadcast kernal is strongly affected by the array's shape.
50*50 is almost 5 times faster than 5*500.
Maybe the `exclude` in `preprocess` should be omitted in this case for better performance?

N5N3 committed:
Merge branch 'JuliaLang:master' into master

oscardssmith commented:
Why did you close this?

N5N3 committed:
revert the a .= b branch

N5N3 commented:
Just for reverting the branch, and try to speed up `copyto_unalias!`.
The acceleration is considerable on my Desktop, but the code might be unclear (include some `@simd` that should not appear)

N5N3 committed:
a faster copyto_unaliased

N5N3 commented:
Opps, it seems we can't use `@simd` here. But it does speed up a lot.

N5N3 committed:
revert

N5N3 committed:
Update broadcast.jl

N5N3 committed:
Merge pull request #1 from N5N3/patch__

revert

stevengj commented on base/broadcast.jl:
```diff
@@ -991,11 +991,17 @@ preprocess_args(dest, args::Tuple{}) = ()
     axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
-    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match
+    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match.
+    # However copyto!(dest, A) is very slow in many cases, implement a faster version here.
```
Wouldn't it be better to speed up `copyto!(dest, A)`?

N5N3 replied:
In fact, I tried, but the problem seems to be a little complicated with the following facts:

1. `copyto!(::Array,::Array)` call C's `memmove`, but it seems slower than a single loop:
```julia
julia> @btime copyto!($b,$a);                    # this call C's memmove
  120.524 ns (0 allocations: 0 bytes)
julia> @btime copyto!(IndexLinear(),$b,IndexLinear(),$a);  # this call Julia's copyto_unalias!
  74.512 ns (0 allocations: 0 bytes)
```
2.  For `abstractarray`, I can't find a way to accelerate `copyto_unalias!(dest, A)` without `@simd`, as:
```julia
a = randn(40,40); b = similar(a);
f1!(a,b) = @inbounds for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end

f2!(a,b) = @inbounds @simd for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end
@btime f1!($a, $b)   #247.132 ns (0 allocations: 0 bytes)
@btime f2!($a, $b)  #220.548 ns (0 allocations: 0 bytes)  a little faster
@btime @views f1!($a[1:end,1:end], $b[1:end,1:end]) #   1.710 μs (0 allocations: 0 bytes)
@btime @views f2!($a[1:end,1:end], $b[1:end,1:end]) #   262.500 ns (0 allocations: 0 bytes) much faster
```

stevengj replied:
Why can't you use `@simd`?

N5N3 replied:
The buildbot throw an error that ’@simd‘ is not defined. I‘m not familiar with the build system, but it seems to be a world age problem?(abstractarray.jl is included before simd.jl in Bass.jl)

stevengj replied:
Sounds just like a bootstrapping problem, i.e. `@simd` is not defined until later in the build process.  That should be fixable by e.g. by moving the `copyto!` definition to a file loaded later.

N5N3 committed:
Revert "revert"

N5N3 committed:
Update Base.jl

Put SimdLoop in advance to avoid bootstrapping (I hope this work)

N5N3 committed:
Merge pull request #2 from N5N3/revert-1-patch__

Focus on copyto_unalias!, put SimdLoop in advance.

N5N3 committed:
Update compiler.jl

N5N3 committed:
Update compiler.jl

N5N3 committed:
Update Base.jl

N5N3 committed:
Update abstractarray.jl

Move some copyto! definitions to other place.

N5N3 committed:
Create abstractarraypatch.jl

These definitions need @simd.

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update Base.jl

N5N3 committed:
Update abstractarraypatch.jl

add license info

N5N3 committed:
Merge pull request #3 from N5N3/patch2

second try

N5N3 committed:
Update abstractarraypatch.jl

N5N3 committed:
Create abstractarraypatch.jl

use view() instead of @view

N5N3 committed:
Update abstractarraypatch.jl

add a branch to speed up Line To Cart which have the same length

N5N3 committed:
Update abstractarraypatch.jl

Use manually expanded SIMD kernal instead of "illegal" `@simd`.
Linear to Cartesian copy is even faster when length(dest) > length(src)

mcabbott commented:
See also #38014, a less elaborate attempt which ran into I think the same bootstrap problem. 

There were concerns raised there about arrays like `BitArray` which alias themselves; I don't know whether or not this PR avoids that.

KristofferC commented:
> There were concerns raised there about arrays like BitArray which alias themselves; 

That's only relevant if one uses the `ivdep` though? I'm a bit surprised why `@simd` should have any effect in copying. I don't think it should give any extra leeway to the compiler so not sure why it would improve performance.

N5N3 committed:
Update abstractarraypatch.jl

Fix for L2Ccopyto!

stevengj commented:
Yes, I can't see any performance difference from `@simd` in something like:
```jl
function foo!(dest, src)
    for i in eachindex(src)
        @inbounds dest[i] = src[i]
    end
end
```
which already uses `vmovups` according to `@code_native`.   What is the case that benefits from `@simd`?

vtjnash commented:
IIRC, the `@simd` macro also tries to rearrange some loops to attempt to make them more linear and more amenable to analysis, though you can get the same (or better?) just by not forcing it to use `IndexCartesian()`

N5N3 commented:
I hope this is reproducable @stevengj 
```julia
function simdfoo!(dest, src)
    @inbounds @simd for i in eachindex(src)
        dest[i] = src[i]
    end
end
function foo!(dest, src)
    @inbounds for i in eachindex(src)
        dest[i] = src[i]
    end
end

a = randn(8,8,8,8);
a = @view a[axes(a)...];
b = similar(a);
b = @view b[axes(a)...];
@btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
@btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
```

KristofferC commented:
I think this is https://github.com/JuliaLang/julia/issues/38073 cc @kimikage 

On 1.5:

```
julia> @btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
  1.658 μs (0 allocations: 0 bytes)

julia> @btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
  2.368 μs (0 allocations: 0 bytes)
```

On 1.6:

```
julia> @btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
  6.112 μs (0 allocations: 0 bytes)

julia> @btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
  1.768 μs (0 allocations: 0 bytes)
```

N5N3 commented:
Some more benchmarks, It seems 1.6.1 is stably slower than 1.5.4 even with a `@simd` on 4，5，6D `CartesianIndices`'s iteration .
```julia
julia> a = randn(4,4,4,4,4,4); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 9.200 μs      1.5.4: 3.112 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 3.763 μs      1.5.4: 6.040 μs

julia> a = randn(5,5,5,5,5); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 6.000 μs      1.5.4: 2.144 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 2.378 μs      1.5.4: 3.287 μs

julia> a = randn(8,8,8,8); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 6.420 μs      1.5.4: 1.840 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 1.990 μs      1.5.4: 2.678 μs

julia> a = randn(16,16,16); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 5.833 μs      1.5.4: 961.857 ns
julia> @btime simdfoo!($b,$a);        #  1.6.1: 747.619 ns    1.5.4: 999.900 ns

julia> a = randn(64,64); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 3.737 μs      1.5.4: 5.650 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 512.500 ns    1.5.4: 699.306 ns
```

N5N3 commented:
@mcabbott 
> There were concerns raised there about arrays like `BitArray` which alias themselves; I don't know whether or not this PR avoids that. (E.g. I don't know whether messing with `Expr(:loopinfo` is identical to `@simd`.)

I tried to remove every `Expr(:loopinfo` generated by `@simd` while keeping the loop body.
The results shows that the acceleration is not influenced. 
Since the expanded code is equivalent, I think this is safe enough?
Maybe we can add a safe option for `@simd` like:
```julia
@simd safe for i in 1:10
end
```
which only expands the loop but not marks it as SIMD.

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update abstractarray.jl

N5N3 committed:
Update abstractarray.jl

N5N3 committed:
Update multidimensional.jl

N5N3 committed:
Merge pull request #4 from N5N3/patch2

remove `@simd`

N5N3 committed:
Update multidimensional.jl

N5N3 committed:
Delete abstractarraypatch.jl

N5N3 committed:
Update Base.jl

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update multidimensional.jl

fix for 0d Cartesian AbstractArray. 
This version should be fast enough if the size of Cartesian array's first dim is larger than 16 (eltype Float64).

N5N3 committed:
Merge pull request #6 from N5N3/patch2

fix for 0d Cartesian AbstractArray. 
This version should be fast enough if the size of Cartesian array's first dim is larger than 16 (eltype Float64).

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update multidimensional.jl

white space

N5N3 committed:
Update abstractarray.jl

white space

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update multidimensional.jl

N5N3 committed:
Update multidimensional.jl

Fix for other IndexStyle. Only use manually expanded version when the size of 1st dim >=16

N5N3 committed:
Merge pull request #7 from N5N3/patch2

Patch2

N5N3 committed:
Merge branch 'JuliaLang:master' into patch2

N5N3 committed:
Update multidimensional.jl

fix typo error.

N5N3 committed:
Update copy.jl

add test for expanded version.

N5N3 committed:
Update multidimensional.jl

white space

N5N3 committed:
Merge pull request #8 from N5N3/patch2

fix white space, typo error; add test;
